internal/executor/instance/persistentworker/isolation/tart/tart.go

```diff
@@ -78,38 +85,100 @@ func New(
 	return tart, nil
 }
 
-func (tart *Tart) Run(ctx context.Context, config *runconfig.RunConfig) (err error) {
+func NewFromIsolation(iso *api.Isolation_Tart_, opts ...Option) (*Tart, error) {
+	tart, err := New(iso.Tart.Image, iso.Tart.User, iso.Tart.Password, iso.Tart.Cpu, iso.Tart.Memory,
+		opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	tart.diskSize = iso.Tart.DiskSize
+	tart.display = iso.Tart.Display
+	tart.volumes = iso.Tart.Volumes
+
+	tart.isolation = &api.Isolation{
+		Type: iso,
+	}
+
+	tart.mountTemporaryWorkingDirectoryFromHost = iso.Tart.MountTemporaryWorkingDirectoryFromHost
+
+	return tart, nil
+}
+
+func (tart *Tart) Pull(ctx context.Context, env map[string]string, logger *echelon.Logger) error {
+	_, _, err := CmdWithLogger(ctx, env, logger, "pull", tart.vmName)
+
+	return err
+}
+
+func (tart *Tart) FQN(ctx context.Context) (string, error) {
+	stdout, _, err := Cmd(ctx, nil, "fqn", tart.vmName)
+	if err != nil {
+		return "", err
+	}
+
+	fqn := strings.TrimSpace(stdout)
+
+	if fqn == "" {
+		return "", fmt.Errorf("%w from Tart", abstract.ErrEmptyFQN)
+	}
+
+	return fqn, nil
+}
+
+//nolint:gocognit // keep it the same as it was before for now to aid in PR review
+func (tart *Tart) CloneConfigureStart(
+	ctx context.Context,
+	config *runconfig.RunConfig,
+) (*abstract.CloneAndConfigureResult, error) {
 	ctx, prepareInstanceSpan := tracer.Start(ctx, "prepare-instance")
 	defer prepareInstanceSpan.End()
 
+	if cloneAndConfigureResult := tart.cloneAndConfigureResult.Load(); cloneAndConfigureResult != nil {
+		return cloneAndConfigureResult, nil
+	}
```

Comment:
I wonder if things down below should be synchronized. Can `CloneConfigureStart` be called twice? It seems it can due to async nature of the current usage of `Slot#Start`.

Reply:
> I wonder if things down below should be synchronized.

By "things" you mean (1) all of the methods after the `CloneConfigureStart()` like `Run()` or (2) just the code below in the same method?

If it's the latter, a second call to `CloneConfigureStart()` would return the `CloneAndConfigureResult` from the atomic variable instead of proceeding further (so we have a thread-safe method, but non-reentrant).

>Can `CloneConfigureStart` be called twice? It seems it can due to async nature of the current usage of `Slot#Start`.

`CloneConfigureStart` is called from two places currently:

* `Slot.Start()` — calls `CloneConfigureStart()` only once and gets a new instance each time (created in `TryStart()`)
* `Run()` — this method is in turn called from `task.go`, but only after the Persistent Worker gets ahold of the standby instance, which means that the `Slot.Start()` had finished

I don't see how/where it can be called twice.

But again, I could make it synchronized similarly to how we use `sync.Mutex` in `Standby if you think it's better. Should I go for it?